Modules/Tests/YosemiteTests/Tools/POS/POSCatalogPersistenceServiceTests.swift

```diff
@@ -221,4 +221,278 @@ struct POSCatalogPersistenceServiceTests {
+        let product = POSProduct.fake().copy(siteID: sampleSiteID, productID: 1, attributes: [attribute1, attribute2])
+        try await insertProduct(product)
+
+        // When
+        let updatedAttribute1 = attribute1.copy(options: ["Cardinal", "Blue"])
+        let newAttribute = ProductAttribute.fake().copy(name: "Material")
+        let updatedProduct = POSProduct.fake().copy(siteID: sampleSiteID, productID: 1, attributes: [newAttribute, updatedAttribute1])
+        let updateCatalog = POSCatalog(products: [updatedProduct], variations: [])
+        try await sut.persistIncrementalCatalogData(updateCatalog, siteID: sampleSiteID)
+
+        // Then
+        try await grdbManager.databaseConnection.read { db in
+            let attributeCount = try PersistedProductAttribute.fetchCount(db)
+            #expect(attributeCount == 2)
+
+            let attributes = try PersistedProductAttribute.fetchAll(db).sorted(by: { $0.name < $1.name })
+            #expect(attributes[0].name == "Color")
+            #expect(attributes[0].options == ["Cardinal", "Blue"])
+            #expect(attributes[0].productID == 1)
+            #expect(attributes[1].name == "Material")
+            #expect(attributes[1].options == [])
+            #expect(attributes[1].productID == 1)
+        }
+    }
+
+    @Test func persistIncrementalCatalogData_replaces_images_for_updated_product() async throws {
+        // Given
+        let image1 = ProductImage.fake().copy(imageID: 1, src: "https://example.com/image1.jpg")
+        let image2 = ProductImage.fake().copy(imageID: 2, src: "https://example.com/image2.jpg")
+        let product = POSProduct.fake().copy(siteID: sampleSiteID, productID: 1, images: [image2, image1])
+        try await insertProduct(product)
+
+        // When
+        let updatedImage1 = image1.copy(src: "https://example.com/image1-1.jpg")
+        let newImage = ProductImage.fake().copy(imageID: 3, src: "https://example.com/image3.jpg")
+        let updatedProduct = POSProduct.fake().copy(siteID: sampleSiteID, productID: 1, images: [newImage, updatedImage1])
+        let updateCatalog = POSCatalog(products: [updatedProduct], variations: [])
+        try await sut.persistIncrementalCatalogData(updateCatalog, siteID: sampleSiteID)
+
+        // Then
+        try await grdbManager.databaseConnection.read { db in
+            let attributeCount = try PersistedProductImage.fetchCount(db)
+            #expect(attributeCount == 2)
+
+            let attributes = try PersistedProductImage.fetchAll(db).sorted(by: { $0.id < $1.id })
+            #expect(attributes[0].src == "https://example.com/image1-1.jpg")
+            #expect(attributes[0].productID == 1)
+            #expect(attributes[1].src == "https://example.com/image3.jpg")
+            #expect(attributes[1].productID == 1)
+        }
+    }
+
+    @Test func persistIncrementalCatalogData_upserts_existing_and_new_products() async throws {
```

Comment:
Strictly, I don't think it does.

Upsert is `INSERT x ON CONFLICT DO UPDATE SET x.name = name {etc}` – there is an `upsert(db)` function we could use for this, at least for mutable persistable records.

Currently we use `insert(db, onConflict: .replace)`, which generates `INSERT x ON CONFLICT DO REPLACE`.

AIUI, the main difference is the hooks that are triggered. With replace, records are deleted and recreated, so the deletions will cascade. For upsert, they're just updated so none of the deletion hooks run.

Reply:
Good to know about the difference. I renamed the products/variations test cases to mention "replace" instead of "upsert" in https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7b371910f91026ef6044503bbf90f580773bf10a and https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/20b59103986da832531d77e0a66926d784e365ed.